src/Rule/DateTimeHandler.php

```diff
@@ -0,0 +1,45 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Validator\Rule;
+
+use Yiisoft\Validator\Result;
+use Yiisoft\Validator\ValidationContext;
+use Yiisoft\Validator\RuleHandlerInterface;
+use Yiisoft\Validator\Exception\UnexpectedRuleException;
+
+use function is_string;
+
+/**
+ * Validates that the value is a valid date.
+ */
+final class DateTimeHandler implements RuleHandlerInterface
+{
+    public function validate(mixed $value, object $rule, ValidationContext $context): Result
+    {
+        if (!$rule instanceof DateTime) {
+            throw new UnexpectedRuleException(DateTime::class, $rule);
+        }
+
+        $result = new Result();
+
+        if (!is_string($value) || empty($value)) {
+            return $result->addError($rule->getIncorrectInputMessage(), [
+                'attribute' => $context->getTranslatedAttribute(),
+                'type' => get_debug_type($value),
+            ]);
+        }
+
+        $date = \DateTime::createFromFormat($rule->getFormat(), $value);
+
+        if ($date === false  || ($date->format($rule->getFormat()) !== $value)) {
```

Comment:
In what cases the check `$date->format($rule->getFormat()) !== $value` is required?

Format for the methods `\DateTime::createFromFormat()` and `\DateTime::format()` can be different.

Format for `\DateTime::createFromFormat()` can contain some special symbols like `?`, `*` and etc
https://www.php.net/manual/en/datetimeimmutable.createfromformat.php

Reply:
> In what cases the check `$date->format($rule->getFormat()) !== $value` is required?

results may vary

```php
$date = \DateTime::createFromFormat('Y-m-d', '2024-02-33');

echo $date->format('Y-m-d'); // 2024-03-04
```

> 
> Format for the methods `\DateTime::createFromFormat()` and `\DateTime::format()` can be different.
> 
> Format for `\DateTime::createFromFormat()` can contain some special symbols like `?`, `*` and etc https://www.php.net/manual/en/datetimeimmutable.createfromformat.php

Thanks, I didn't think about that.
  What do you think about it?
need to handle error types?

```php
 if ( \DateTime::getLastErrors()) {
```

Reply:
> ```
>  if ( \DateTime::getLastErrors()) {
> ```

Looks like good solution instead of `$date->format($rule->getFormat()) !== $value`